src/cid.rs

```diff
@@ -91,34 +91,57 @@ where
     }
 
     fn to_string_v1(&self) -> String {
-        multibase::encode(Base::Base32Lower, self.to_bytes().as_slice())
+        multibase::encode(Base::Base32Lower, &self.to_bytes())
     }
 
-    fn to_bytes_v0(&self) -> Vec<u8> {
-        self.hash.to_vec()
+    fn copy_to_slice_v0(&self, buffer: &mut [u8]) -> usize {
```

Comment:
If we want such a signature is should use [`Read`](https://doc.rust-lang.org/std/io/trait.Read.html) and not a custom method.

Reply:
Write only works in a std environment, it's not the same as `copy_to_slice`. Write on a vec will extend the backing storage if it's out of bounds. `GenericArray` doesn't help as you don't know the size of the cid before hand.

Reply:
> GenericArray doesn't help as you don't know the size of the cid before hand.

But you do, if the multihash is stored as one too

Reply:
> Write on a vec will extend the backing storage if it's out of bounds.

That is true, I guess I am not yet grasping when and how you actually plan to use this method.

Reply:
to use stack allocation you need to know at compile time. so you'd have to allocate the maximum size it could be.

Reply:
Not that there is anything wrong with that

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f6522eac7f941309b887b5593627dfa5 shows roughly how to make it work fully static

Reply:
This is how the dag cbor codec uses it
```rust
impl Encode<DagCbor> for Cid {
    #[inline]
    fn encode<W: Write>(&self, w: &mut W) -> Result<()> {
        write_tag(w, 42)?;
        // insert zero byte per https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-cbor.md#links
        // 128 bytes should be plenty large enough for any cid
        let mut buf = [0u8; 128];
        let len = self.copy_to_slice(&mut buf);
        write_u64(w, 2, len as u64 + 1)?;
        w.write_all(&[0])?;
        w.write_all(&buf[..len])?;
        Ok(())
    }
}
```

Reply:
I guess it can just aswell take a writer

Reply:
So version and codec are varints, you'd have three Size parameters?